### PR TITLE
registry-facade: retry on io.CopyBuffer fails

### LINKED ISSUE #14264: [registry-facade] Error on workspace start: `cannot pull image - failed to pull and unpack image` aka unexpected EOF
## Bug description

We have observed the following error up to 132 times in 24 hours:

> cannot pull image: rpc error: code = Unknown desc = failed to pull and unpack image “reg.ws-eu73.gitpod.io:20000/remote/7d932e9e-914c-4c14-8341-6d04f079b254:latest”: failed to read expected number of bytes: unexpected EOF

The error did not recur on user retry.

context `https://gitpod.io/#https://github.com/gitpod-io/website/pull/2960`

## Steps to reproduce

N/A

## Workspace affected

https://gitpodio-website-u2l4lu0fymu.ws-eu73.gitpod.io/

## Plan

- [x] Add additional logs to more easily identify the source of the problem
- [ ] Fix failure of IPFS blobsource, as described by @csweichel [[1](https://github.com/gitpod-io/gitpod/issues/14264#issuecomment-1326133073)]

## Anything else

We now have new logs and recent occurrences in gen75 to facilitate the investigation [[1](https://cloudlogging.app.goo.gl/JB82pNWbk5s5saHt9)].

This also was happening in gen72 and gen73 according to [logs](https://cloudlogging.app.goo.gl/T8ozTgZJDxDrMwUU9).

image

### PR DESCRIPTION
**Pull request by jenting**
## Description

Retry on io.CopyBuffer fails when it is either `connection reset by peer` or `broken pipe` error.
With a retry interval of 100ms and a maximum retry time of 1 minute.

## Related Issue(s)

Fixes #14264

## How to test

Check the unit test.

## Release Notes

```release-note
None
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/registry-facade/pkg/registry/blob.go [hunk 1]
```diff
@@ -171,16 +171,27 @@ func (bh *blobHandler) getBlob(w http.ResponseWriter, r *http.Request) {
...
 
-		n, err := io.CopyBuffer(w, rc, *bp)
+		const retryMax int = 5
+		var n int64
+		for i := 0; i <= retryMax; i++ {
+			n, err = io.CopyBuffer(w, rc, *bp)
+			if err == nil {
+				break
+			}
+		}
```
Comment (utam0k) [inline components/registry-facade/pkg/registry/blob.go:181]: I have some questions about the retry mechanism.
- Is there a reason that we retry any error? In other words, I wonder if we have to check what kind of error.
- Should we put the sleep before next challenge?

Comment (jenting) [reply components/registry-facade/pkg/registry/blob.go]: In the last 7 days [log](https://cloudlogging.app.goo.gl/kHD1XC9YRsHkKwRf8), the errors are `connection reset by peer` and `broken pipe`. We should retry with these two errors only.

**Review by utam0k — APPROVED** [review]

**Review by mrsimonemms — APPROVED** [review]

**Review by Pothulapati — APPROVED** [review]
LGTM

**Review by felladrin — APPROVED** [review]

**Pull request merged**